RevWalkException in RefWithCommitIterator

## Expected Behavior

Engine should be able to analyze PGA.

## Current Behavior

It fails with exception:
```
Exception in thread "main" org.apache.spark.SparkException: Job aborted due to stage failure: Task 92 in stage 1.0 failed 4 times, most recent failure: Lost task 92.3 in stage 1.0 (TID 530, 10.2.101.146, executor 7): org.eclipse.jgit.errors.RevWalkException: Walk failure.
	at org.eclipse.jgit.revwalk.RevWalk.iterator(RevWalk.java:1331)
	at scala.collection.convert.Wrappers$JIterableWrapper.iterator(Wrappers.scala:54)
	at scala.collection.IterableLike$class.toIterator(IterableLike.scala:104)
	at scala.collection.AbstractIterable.toIterator(Iterable.scala:54)
	at tech.sourced.engine.iterator.RefWithCommitIterator.liftedTree1$1(CommitIterator.scala:141)
	at tech.sourced.engine.iterator.RefWithCommitIterator.hasNext(CommitIterator.scala:138)
	at tech.sourced.engine.iterator.ChainableIterator.hasNext(ChainableIterator.scala:89)
	at scala.collection.Iterator$class.isEmpty(Iterator.scala:330)
	at tech.sourced.engine.iterator.ChainableIterator.isEmpty(ChainableIterator.scala:16)
	at tech.sourced.engine.iterator.ChainableIterator.hasNext(ChainableIterator.scala:90)
	at scala.collection.Iterator$class.isEmpty(Iterator.scala:330)
	at tech.sourced.engine.iterator.ChainableIterator.isEmpty(ChainableIterator.scala:16)
	at tech.sourced.engine.iterator.ChainableIterator.hasNext(ChainableIterator.scala:90)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at tech.sourced.engine.iterator.CleanupIterator.hasNext(CleanupIterator.scala:23)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at com.datastax.spark.connector.util.CountingIterator.hasNext(CountingIterator.scala:12)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.next(GroupingBatchBuilder.scala:105)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.next(GroupingBatchBuilder.scala:31)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.foreach(GroupingBatchBuilder.scala:31)
	at com.datastax.spark.connector.writer.TableWriter$$anonfun$writeInternal$1.apply(TableWriter.scala:233)
	at com.datastax.spark.connector.writer.TableWriter$$anonfun$writeInternal$1.apply(TableWriter.scala:210)
	at com.datastax.spark.connector.cql.CassandraConnector$$anonfun$withSessionDo$1.apply(CassandraConnector.scala:112)
	at com.datastax.spark.connector.cql.CassandraConnector$$anonfun$withSessionDo$1.apply(CassandraConnector.scala:111)
	at com.datastax.spark.connector.cql.CassandraConnector.closeResourceAfterUse(CassandraConnector.scala:145)
	at com.datastax.spark.connector.cql.CassandraConnector.withSessionDo(CassandraConnector.scala:111)
	at com.datastax.spark.connector.writer.TableWriter.writeInternal(TableWriter.scala:210)
	at com.datastax.spark.connector.writer.TableWriter.insert(TableWriter.scala:197)
	at com.datastax.spark.connector.writer.TableWriter.write(TableWriter.scala:183)
	at com.datastax.spark.connector.RDDFunctions$$anonfun$saveToCassandra$1.apply(RDDFunctions.scala:36)
	at com.datastax.spark.connector.RDDFunctions$$anonfun$saveToCassandra$1.apply(RDDFunctions.scala:36)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

Log from worker: 
[stderr_worker.txt](https://github.com/src-d/engine/files/1767048/stderr_worker.txt)


## Steps to Reproduce (for bugs)

```
spark-shell --packages "tech.sourced:engine:0.5.2" --master="spark://p-spark-master:7077"

scala> import tech.sourced.engine._
import tech.sourced.engine._

scala> val engine = Engine(spark, "hdfs://hdfs-namenode/apps/borges/10k", "siva")
engine: tech.sourced.engine.Engine = tech.sourced.engine.Engine@4086e003

scala> engine.getRepositories.getHEAD.getCommits.getTreeEntries.count()
<fail>
```

## Context
running gemini

## Your Environment (for bugs)

* Version used: 0.5.2
* Src-d staging cluster

Following @bzz's suggestion here https://github.com/src-d/gemini/issues/42#issuecomment-372258962, I tried to add a catch for `RevWalkException` in `RefWithCommitIterator.hasNext`. That solves the original error reported by @smacker, but I was still getting this exception somewhere else:

```
WARN TaskSetManager: Lost task 81.0 in stage 0.0 (TID 83, 10.2.54.3, executor 1): org.eclipse.jgit.errors.RevWalkException: Walk failure.
	at org.eclipse.jgit.revwalk.RevWalk$3.next(RevWalk.java:1353)
	at org.eclipse.jgit.revwalk.RevWalk$3.next(RevWalk.java:1338)
	at scala.collection.convert.Wrappers$JIteratorWrapper.next(Wrappers.scala:43)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:393)
	at tech.sourced.engine.iterator.RefWithCommitIterator.next(CommitIterator.scala:165)
	at tech.sourced.engine.iterator.RefWithCommitIterator.next(CommitIterator.scala:123)
	at tech.sourced.engine.iterator.ChainableIterator.nextRaw(ChainableIterator.scala:130)
	at tech.sourced.engine.iterator.ChainableIterator.hasNext(ChainableIterator.scala:97)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at tech.sourced.engine.iterator.CleanupIterator.hasNext(CleanupIterator.scala:23)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.agg_doAggregateWithoutKey$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:125)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.eclipse.jgit.errors.MissingObjectException: Missing commit 9ef24c5f907acedada6858991500d9dfc8b9456a
	at org.eclipse.jgit.internal.storage.file.WindowCursor.open(WindowCursor.java:164)
	at org.eclipse.jgit.revwalk.RevWalk.getCachedBytes(RevWalk.java:903)
	at org.eclipse.jgit.revwalk.RevCommit.parseHeaders(RevCommit.java:155)
	at org.eclipse.jgit.revwalk.PendingGenerator.next(PendingGenerator.java:147)
	at org.eclipse.jgit.revwalk.RevWalk.next(RevWalk.java:435)
	at org.eclipse.jgit.revwalk.RevWalk$3.next(RevWalk.java:1350)
	... 24 more
```

I added a try/catch to `ChainableIterator.hasNext`, unfortunately I had to remove the `@tailrec` optimization because of it.

You can see the PR here . I hope it's useful, but please feel free to discard it and implement it in a proper way.

@bzz sorry for the line numbers confusion, I had some extra lines with logs to figure out what was going on. I've reproduced the experiment as you requested.

Changes: One commit in a new [branch of my fork](https://github.com/carlosms/engine/tree/issue-336-test): https://github.com/carlosms/engine/commit/f21ab65f612bc6504cf31b3ef16aaf53b0df5b70

The relevant failure is below, and you can also see the spark-shell output [in this gist](https://gist.github.com/carlosms/34783ab819790fbb6a530899b6ca7560).

```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 81 in stage 0.0 failed 4 times, most recent failure: Lost task 81.3 in stage 0.0 (TID 215, 10.2.61.3, executor 1): org.eclipse.jgit.errors.RevWalkException: Walk failure.
	at org.eclipse.jgit.revwalk.RevWalk$3.next(RevWalk.java:1353)
	at org.eclipse.jgit.revwalk.RevWalk$3.next(RevWalk.java:1338)
	at scala.collection.convert.Wrappers$JIteratorWrapper.next(Wrappers.scala:43)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:393)
	at tech.sourced.engine.iterator.RefWithCommitIterator.next(CommitIterator.scala:168)
	at tech.sourced.engine.iterator.RefWithCommitIterator.next(CommitIterator.scala:127)
	at tech.sourced.engine.iterator.ChainableIterator.nextRaw(ChainableIterator.scala:120)
	at tech.sourced.engine.iterator.ChainableIterator.hasNext(ChainableIterator.scala:94)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at tech.sourced.engine.iterator.CleanupIterator.hasNext(CleanupIterator.scala:23)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.agg_doAggregateWithoutKey$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:125)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.eclipse.jgit.errors.MissingObjectException: Missing commit 9ef24c5f907acedada6858991500d9dfc8b9456a
	at org.eclipse.jgit.internal.storage.file.WindowCursor.open(WindowCursor.java:164)
	at org.eclipse.jgit.revwalk.RevWalk.getCachedBytes(RevWalk.java:903)
	at org.eclipse.jgit.revwalk.RevCommit.parseHeaders(RevCommit.java:155)
	at org.eclipse.jgit.revwalk.PendingGenerator.next(PendingGenerator.java:147)
	at org.eclipse.jgit.revwalk.RevWalk.next(RevWalk.java:435)
	at org.eclipse.jgit.revwalk.RevWalk$3.next(RevWalk.java:1350)
	... 24 more

Driver stacktrace:
  at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1499)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1487)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1486)
  at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
  at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1486)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:814)
  at scala.Option.foreach(Option.scala:257)
  at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:814)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1714)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1669)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1658)
  at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
  at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:630)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2022)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2043)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2062)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:2087)
  at org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:936)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
  at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
  at org.apache.spark.rdd.RDD.collect(RDD.scala:935)
  at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:278)
  at org.apache.spark.sql.Dataset$$anonfun$count$1.apply(Dataset.scala:2430)
  at org.apache.spark.sql.Dataset$$anonfun$count$1.apply(Dataset.scala:2429)
  at org.apache.spark.sql.Dataset$$anonfun$55.apply(Dataset.scala:2837)
  at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:65)
  at org.apache.spark.sql.Dataset.withAction(Dataset.scala:2836)
  at org.apache.spark.sql.Dataset.count(Dataset.scala:2429)
  ... 50 elided
Caused by: org.eclipse.jgit.errors.RevWalkException: Walk failure.
  at org.eclipse.jgit.revwalk.RevWalk$3.next(RevWalk.java:1353)
  at org.eclipse.jgit.revwalk.RevWalk$3.next(RevWalk.java:1338)
  at scala.collection.convert.Wrappers$JIteratorWrapper.next(Wrappers.scala:43)
  at scala.collection.Iterator$$anon$10.next(Iterator.scala:393)
  at tech.sourced.engine.iterator.RefWithCommitIterator.next(CommitIterator.scala:168)
  at tech.sourced.engine.iterator.RefWithCommitIterator.next(CommitIterator.scala:127)
  at tech.sourced.engine.iterator.ChainableIterator.nextRaw(ChainableIterator.scala:120)
  at tech.sourced.engine.iterator.ChainableIterator.hasNext(ChainableIterator.scala:94)
  at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
  at tech.sourced.engine.iterator.CleanupIterator.hasNext(CleanupIterator.scala:23)
  at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
  at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
  at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.agg_doAggregateWithoutKey$(Unknown Source)
  at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
  at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
  at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
  at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
  at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:125)
  at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
  at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
  at org.apache.spark.scheduler.Task.run(Task.scala:108)
  at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)
Caused by: org.eclipse.jgit.errors.MissingObjectException: Missing commit 9ef24c5f907acedada6858991500d9dfc8b9456a
  at org.eclipse.jgit.internal.storage.file.WindowCursor.open(WindowCursor.java:164)
  at org.eclipse.jgit.revwalk.RevWalk.getCachedBytes(RevWalk.java:903)
  at org.eclipse.jgit.revwalk.RevCommit.parseHeaders(RevCommit.java:155)
  at org.eclipse.jgit.revwalk.PendingGenerator.next(PendingGenerator.java:147)
  at org.eclipse.jgit.revwalk.RevWalk.next(RevWalk.java:435)
  at org.eclipse.jgit.revwalk.RevWalk$3.next(RevWalk.java:1350)
  ... 24 more
```

I'm still trying to figure out how to find the siva file

--- a/src/main/scala/tech/sourced/engine/iterator/CommitIterator.scala
+++ b/src/main/scala/tech/sourced/engine/iterator/CommitIterator.scala
@@ -4,11 +4,16 @@ import java.sql.Timestamp
 
 import org.apache.spark.internal.Logging
 import org.eclipse.jgit.api.Git
-import org.eclipse.jgit.errors.{IncorrectObjectTypeException, MissingObjectException}
+import org.eclipse.jgit.errors.{
+  IncorrectObjectTypeException,
+  MissingObjectException,
+  RevWalkException
+}
 import org.eclipse.jgit.lib.{ObjectId, Ref, Repository}
 import org.eclipse.jgit.revwalk.RevCommit
 import tech.sourced.engine.util.{CompiledFilter, Filters}
 
+import scala.annotation.tailrec
 import scala.collection.JavaConverters.iterableAsScalaIterableConverter
 
 /**
@@ -120,7 +125,7 @@ object CommitIterator {
   * @param refs       iterator of references
   * @param maxResults max results to return
   */
-class RefWithCommitIterator(repo: Repository,
+private[iterator] class RefWithCommitIterator(repo: Repository,
                             refs: Iterator[Ref],
                             maxResults: Int = 0
                            ) extends Iterator[ReferenceWithCommit] with Logging {
@@ -128,9 +133,17 @@ class RefWithCommitIterator(repo: Repository,
   private var actualRef: Ref = _
   private var commits: Iterator[RevCommit] = _
   private var index: Int = 0
+  private var nextResult: ReferenceWithCommit = _
+  private var consumed: Int = 0
 
   /** @inheritdoc */
-  override def hasNext: Boolean = {
+  @tailrec
+  final override def hasNext: Boolean = {
+    // do not advance the iterator until the next result has been consumed
+    if (nextResult != null) {
+      return true
+    }
+
     while ((commits == null || !commits.hasNext) && refs.hasNext) {
       actualRef = refs.next()
       index = 0
@@ -147,19 +160,42 @@ class RefWithCommitIterator(repo: Repository,
             log.warn("missing object", e)
             null
         }
-
-      if (maxResults > 0 && commits != null) {
-        commits = commits.take(maxResults)
-      }
     }
 
-    refs.hasNext || (commits != null && commits.hasNext)
+
+    if (maxResults > 0 && consumed == maxResults) {
+      false
+    } else  if (refs.hasNext || (commits != null && commits.hasNext)) {
+      // Preload the result, so if there's an exception we can skip it and
+      // continue with the next. If we don't preload, the exception will
+      // happen in next, and that's an odd behaviour. This behaviour should
+      // be completely transparent to the iterator user.
+      try {
+        nextResult = ReferenceWithCommit(actualRef, commits.next(), index)
+        index += 1
+        true
+      } catch {
+        case e: RevWalkException =>
+          log.warn("rev walk error", e)
+          this.hasNext
+      }
+    } else {
+      false
+    }
   }
 
-  /** @inheritdoc */
+  /**
+    * Returns the next [[ReferenceWithCommit]]. It should never be called after
+    * hasNext has returned "false", since this can lead to incorrect results.
+    * It can only be called after a call to hasNext. Repeatedly calling this method
+    * will result in an error.
+    *
+    * @return a reference with commit
+    */
   override def next(): ReferenceWithCommit = {
-    val result: ReferenceWithCommit = ReferenceWithCommit(actualRef, commits.next(), index)
-    index += 1
+    val result = nextResult
+    nextResult = null
+    consumed += 1
     result
   }
 

